Benp/capture duration times

benpatterson opened the pull request:
work in progress at this point.

Example output:

>
[{'repo': u'edx-app-ios', 'average duration': 65}, {'repo': u'django-oscar', 'average duration': 65}, {'repo': u'bok-choy', 'average duration': 25}, {'repo': u'edx-app-android', 'average duration': 21}, {'repo': u'configuration', 'average duration': 20}, {'repo': u'edx-analytics-dashboard', 'average duration': 16}, {'repo': u'ecommerce', 'average duration': 13}, {'repo': u'edx-notifications', 'average duration': 9}, {'repo': u'edx-ora2', 'average duration': 8}, {'repo': u'edx-ora', 'average duration': 8}, {'repo': u'cs_comments_service', 'average duration': 7}, {'repo': u'edx-drf-extensions', 'average duration': 7}, {'repo': u'edx-documentation', 'average duration': 6}, {'repo': u'auth-backends', 'average duration': 6}, {'repo': u'locust', 'average duration': 6}, {'repo': u'ml-service-api', 'average duration': 6}, {'repo': u'course-discovery', 'average duration': 5}, {'repo': u'ux-pattern-library', 'average duration': 5}, {'repo': u'edx-ui-toolkit', 'average duration': 5}, {'repo': u'diff-cover', 'average duration': 4}, {'repo': u'edx-notes-api', 'average duration': 4}, {'repo': u'ease', 'average duration': 4}, {'repo': u'edx-user-state-client', 'average duration': 4}, {'repo': u'django-oscar-extensions', 'average duration': 4}, {'repo': u'edx-proctoring', 'average duration': 3}, {'repo': u'edx-django-release-util', 'average duration': 3}, {'repo': u'jenkins-job-dsl', 'average duration': 3}, {'repo': u'nose', 'average duration': 3}, {'repo': u'xblock-utils', 'average duration': 3}, {'repo': u'js-test-tool', 'average duration': 3}, {'repo': u'credentials', 'average duration': 3}, {'repo': u'xblock-sdk', 'average duration': 3}, {'repo': u'XBlock', 'average duration': 3}, {'repo': u'discern', 'average duration': 3}, {'repo': u'pa11ycrawler', 'average duration': 2}, {'repo': u'eslint-config-edx', 'average duration': 2}, {'repo': u'edx-analytics-data-api', 'average duration': 2}, {'repo': u'edx-e2e-tests', 'average duration': 2}, {'repo': u'edx-reverification-block', 'average duration': 2}, {'repo': u'opaque-keys', 'average duration': 2}, {'repo': u'edx-lint', 'average duration': 2}, {'repo': u'edx-django-extensions', 'average duration': 2}, {'repo': u'xblock-lti-consumer', 'average duration': 2}, {'repo': u'api-manager', 'average duration': 1}, {'repo': u'xqueue', 'average duration': 1}, {'repo': u'xsy', 'average duration': 1}, {'repo': u'edx-val', 'average duration': 1}, {'repo': u'marketing-site', 'average duration': 1}, {'repo': u'programs', 'average duration': 1}, {'repo': u'cookiecutter-django-ida', 'average duration': 1}, {'repo': u'harprofiler', 'average duration': 1}, {'repo': u'edx-certificates', 'average duration': 1}, {'repo': u'openedx-webhooks', 'average duration': 1}, {'repo': u'edx-analytics-data-api-client', 'average duration': 1}, {'repo': u'edx-organizations', 'average duration': 1}, {'repo': u'edx-search', 'average duration': 1}, {'repo': u'i18n-tools', 'average duration': 1}, {'repo': u'edx-custom-a11y-rules', 'average duration': 1}, {'repo': u'edx-django-sites-extensions', 'average duration': 1}, {'repo': u'django-oauth2-provider', 'average duration': 0}, {'repo': u'tubular', 'average duration': 0}, {'repo': u'edx-gomatic', 'average duration': 0}, {'repo': u'testeng-ci', 'average duration': 0}, {'repo': u'ecommerce-worker', 'average duration': 0}, {'repo': u'edx-load-tests', 'average duration': 0}, {'repo': u'edx-rest-api-client', 'average duration': 0}, {'repo': u'discussions', 'average duration': 0}, {'repo': u'edx-submissions', 'average duration': 0}, {'repo': u'build-pipeline', 'average duration': 0}, {'repo': u'xqueue-watcher', 'average duration': 0}, {'repo': u'event-tracking', 'average duration': 0}, {'repo': u'edx-milestones', 'average duration': 0}, {'repo': u'edx-oauth2-provider', 'average duration': 0}, {'repo': u'repo-tools', 'average duration': 0}, {'repo': u'notifier', 'average duration': 0}, {'repo': u'ccx-keys', 'average duration': 0}, {'repo': u'dummy-webapp', 'average duration': 0}, {'repo': u'edx-sitespeed', 'average duration': 0}]

benpatterson commented:
@cpennington @jzoldak @estute 

Here's the PR I mentioned this morning. Ultimately I want an easy way for us to know which builds are the longest.

A previous comment on this PR gives you example output. (Slight difference is that we will now see it as an INFO message in the current version of code).

Functionally as of this writing:
* It will take the last 5 successful builds of all repos and produce an average number for the duration, per repo
* Note that Travis "duration" is only from when the build actually started to when it completed (it is not when the build was _queued_ to when it finished). 
* In Travis's model, "build" is a parent to "job" which means if repos have multiple sub-jobs (e.g., the configuration repo), this script will tell us how long it takes all of them to complete, on average.
* The script does not look at a specific branch. I thought for now that would be fine but would be open to feedback.
* It is reporting in minutes. 
* The average will report a 0 if the average is under a minute. I could make it always report 1 as the floor if that helps.

Anyway that is essentially what this is producing. I'd be curious your high-level feedback as well as any specific coding feedback.

cpennington commented:
Any reason to truncate to integer minute values, rather than allowing fractional ones?

cpennington commented:
Also, not sure it's really useful with only 5 builds to work from, but it might be worth computing the standard deviation as well as the average, to understand how variable the builds are.

cpennington commented on travis/build_info.py:
```diff
@@ -228,6 +295,8 @@ def main(raw_args):
     if args.task_class.upper() == 'JOB':
         get_job_counts(org=args.org)
+    elif args.task_class.upper() == 'DURATION':
+        get_average_duration_org(org=args.org)
```
Using `click` might make this whole thing cleaner. You could define a `click.group` for the `build_info` script, and then decorate each report function with `click.command`, and it would wire them together for you.

benpatterson commented:
Re: the integer values and an average... in both cases I was trying to make the data easily consumable. (Mentally, I'm picturing a bar graph.) I think getting more granular can be something to consider if this script ends up proving both useful and also not quite enough.

For click, that's also probably worth a refactor later. I've not used click, but can see in the docs that it's making arguments against argparse.

cpennington commented:
I feel like anything you make a bar-chart in is going to be able to handle either ints or floats. In general, they should also be able to do whiskers (deviation).

benpatterson commented:
I *knew* you were going to say that! Let me think about it...

benpatterson commented:
Based on my read of docs, if we want standard deviation, we need to either install numpy or upgrade to python 3.

benpatterson commented:
Hi @cpennington @jzoldak @estute 

At this point, I'm going to be the only person consuming this functionality. Cale's comments are noted though, as things we can do to further improve this. For now it's a fine enough yardstick to hear in general what builds are taking the longest. I do think we'd see a wide standard deviation on some repos, so getting that later will be useful.

I just added more tests and have 100% coverage of the code I've added. 

I'll squash and merge this as-is.
